NAM/activations.h

```diff
@@ -1,15 +1,125 @@
 #pragma once
 
+#include <string>
 #include <cmath> // expf
+#include <unordered_map>
+#include <Eigen/Dense>
 
 namespace activations
 {
-float relu(float x)
+inline float relu(float x)
 {
-  return x > 0.0f ? x : 0.0f;
+return x > 0.0f ? x : 0.0f;
 };
-float sigmoid(float x)
+
+inline float sigmoid(float x)
+{
+return 1.0f / (1.0f + expf(-x));
+};
+
+inline float hard_tanh(float x)
+{
+const float t = x < -1 ? -1 : x;
+return t > 1 ? 1 : t;
+}
+
+inline float fast_tanh(const float x)
+{
+const float ax = fabsf(x);
+const float x2 = x * x;
+
+return (x * (2.45550750702956f + 2.45550750702956f * ax + (0.893229853513558f + 0.821226666969744f * ax) * x2)
+        / (2.44506634652299f + (2.44506634652299f + x2) * fabsf(x + 0.814642734961073f * x * ax)));
+}
+
+class Activation
+{
+public:
+Activation(){};
+  virtual void apply(Eigen::MatrixXf& matrix)
+  {
+    apply(matrix.data(), matrix.rows() * matrix.cols());
+  }
+  virtual void apply(Eigen::Block<Eigen::MatrixXf> block)
+  {
+    apply(block.data(), block.rows() * block.cols());
+  }
+  virtual void apply(Eigen::Block<Eigen::MatrixXf, -1, -1, true> block)
+  {
+    apply(block.data(), block.rows() * block.cols());
+  }
+  virtual void apply(float* data, long size) {}
```

Comment:
Putting functions which will be called per-sample (or maybe multiple times per-sample depending on the network architecture) behind a virtual interface will likely have a significant performance impact, since it prevents the compiler from inlining these function calls in most cases.

In [RTNeural](https://github.com/jatinchowdhury18/RTNeural), this is a big reason why the "dynamic" API is much slower than the "static" API.

Reply:
I was worried about inlining, too, which is why I made sure to keep everything as block operations. I did not see any slowdown over the existing implementation in my testing, but that could be compiler-specific? I'm testing using VS on Windows.

The override methods definitely make it more convenient to use, but it would be worth giving that up if there is a clearly demonstrable performance gain to be had.